WcaOnRails/app/models/competition.rb

```diff
@@ -622,6 +622,62 @@ def continent
     country ? Continent.find_by_id(country.continentId) : nil
   end
 
+  def psych_sheet_event(event)
+    preferred_format = event.preferred_formats.first
+    sort_clause = (preferred_format.sort_by == :single) ? "-single_rank desc, -average_rank desc, users.name" : "-average_rank desc, -single_rank desc, users.name"
+
+    joinsql = <<-ENDSQL
+      join users on users.id = Preregs.user_id
+      join Countries on Countries.iso2 = users.country_iso2
+      left join RanksSingle on RanksSingle.personId = users.wca_id and RanksSingle.eventId = '#{event.id}'
+      left join RanksAverage on RanksAverage.personId = users.wca_id and RanksAverage.eventId = '#{event.id}'
+    ENDSQL
+
+    selectsql = <<-ENDSQL
+      Preregs.id,
+      users.name select_name,
+      users.wca_id select_wca_id,
+      Preregs.accepted_at,
+      Countries.name select_country,
+      registration_events.event_id,
+      RanksAverage.worldRank average_rank,
+      ifnull(RanksAverage.best, 0) average_best,
+      RanksSingle.worldRank single_rank,
+      ifnull(RanksSingle.best, 0) single_best
+    ENDSQL
+
+    registrations = self.registrations.
+                         accepted.
+                         joins(joinsql).
+                         joins(:registration_events).
+                         where("registration_events.event_id=?", event.id).
+                         order(sort_clause).
+                         select(selectsql)
+
+    prev_registration = nil
+    registrations.each_with_index do |registration, i|
+      registration.tied_previous = false
+      if prev_registration
+        registration.tied_previous = if preferred_format.sort_by == :single
+                                       registration.single_rank == prev_registration.single_rank
+                                     else
+                                       registration.average_rank == prev_registration.average_rank
+                                     end
+        registration.pos = registration.tied_previous ? prev_registration.pos : i + 1
+      else
+        registration.pos = i + 1
+      end
+      has_competed = if preferred_format.sort_by == :single
+                       registration.single_rank
+                     else
+                       registration.average_rank
+                     end
+      registration.pos = nil unless has_competed
+      prev_registration = registration
+    end
```

Comment:
@jonatanklosko For some reason the code you suggested is much slower. It takes around 4 seconds to calculate the 500 positions as opposed to much less than 1 second for the original code. Is this because of the send method? Because of this I might keep it like it is.